openfl/component/ca/ca.py

```diff
@@ -0,0 +1,233 @@
+    url = urls[-1]
+    url = url.replace('https', 'http')
+    name = url.split('/')[-1]
+    logger.info(f'Donwloading {name}')
+    urllib.request.urlretrieve(url, f'{prefix}/{name}')
+    shutil.unpack_archive(f'{prefix}/{name}', f'{prefix}/step')
+
+
+def get_token(name, ca_url, ca_path='.'):
+    """
+    Create authentication token.
+
+    Args:
+        name: common name for following certificate
+                    (aggregator fqdn or collaborator name)
+        ca_url: full url of CA server
+    """
+    ca_path = Path(ca_path)
+    step_config_dir = ca_path / 'step_config'
+    pki_dir = ca_path / 'cert'
+    step, _ = get_bin_names(ca_path)
+    if not step:
+        raise Exception('Step-CA is not installed!\nRun `fx pki install` first')
+
+    priv_json = step_config_dir / 'secrets' / 'priv.json'
+    pass_file = pki_dir / 'pass_file'
+    root_crt = step_config_dir / 'certs' / 'root_ca.crt'
+    try:
+        token = subprocess.check_output(
+            f'{step} ca token {name} '
+            f'--key {priv_json} --root {root_crt} '
+            f'--password-file {pass_file} 'f'--ca-url {ca_url}', shell=True)
+    except subprocess.CalledProcessError as exc:
+        logger.error(f'Error code {exc.returncode}: {exc.output}')
+        return
+
+    if token[-1:] == b'\n':
+        token = token[:-1]
+    length = len(token)
+    assert(length < 10000)
+    length = str(10000 + length)[-4:]
+    with open(step_config_dir / 'certs' / 'root_ca.crt', mode='rb') as file:
+        root_ca = file.read()
+
+    base64_bytes = base64.b64encode(root_ca)
+    base64_message = base64_bytes.decode('ascii')
+    return str(length) + token.decode('ascii') + base64_message
+
+
+def get_bin_names(ca_path):
+    """Get paths of step binaries."""
+    ca_path = Path(ca_path)
+    step = None
+    step_ca = None
+    if os.path.exists(ca_path / 'step'):
+        dirs = os.listdir(ca_path / 'step')
+        for dir_ in dirs:
+            if 'step_' in dir_:
+                step = ca_path / 'step' / dir_ / 'bin' / 'step'
+            if 'step-ca' in dir_:
+                step_ca = ca_path / 'step' / dir_ / 'bin' / 'step-ca'
+    return step, step_ca
+
+
+def certify(name, cert_path, token_with_cert):
+    """Create a collaborator manager workspace."""
+    os.makedirs(cert_path, exist_ok=True)
+
+    length = int(token_with_cert[:4])
+    token = token_with_cert[4:length + 4]
+    root_ca = token_with_cert[length + 4:]
+    message_bytes = base64.b64decode(root_ca)
+
+    step, _ = get_bin_names('.')
+    if not step:
+        url = 'http://api.github.com/repos/smallstep/cli/releases/latest'
+        download_step_bin(url, 'step_linux', 'amd', prefix='./')
+        step, _ = get_bin_names('.')
+    assert(step)
+
+    # write ca cert to file
+    with open(f'{cert_path}/root_ca.crt', mode='wb') as file:
+        file.write(message_bytes)
+    # request to ca server
+    os.system(f'./{step} ca certificate {name} {cert_path}/{name}.crt '
```

Comment:
`--ca-url` should be added to this command, otherwise it defaults to `localhost:8443` as the address to the step ca server

Reply:
The address for the step ca server is passed in a JWT token

Reply:
You're right. I had a prior version of step ca / step cli installed, so the generated token was pulling the wrong address.